ValveResourceFormat/Resource/ResourceTypes/KV3.cs

```diff
@@ -0,0 +1,592 @@
+            }
+
+            //Catch comments
+            if (c == '/')
+            {
+                stateStack.Push(State.COMMENT);
+                currentString.Clear();
+                currentString.Append(c);
+                return;
+            }
+
+            //Check for the end of the structure
+            if (c == '}')
+            {
+                KVObject value = objStack.Pop();
+                objStack.Peek().AddProperty(value.key, new KVValue(typeof(KVObject), value));
+                stateStack.Pop();
+                return;
+            }
+
+            //Start looking for the next property name
+            stateStack.Push(State.PROP_NAME);
+            currentString.Clear();
+            currentString.Append(c);
+        }
+        
+        //Read a string value
+        private static void ReadValueString(char c)
+        {
+            if (c == '"' && fileStream[index-1] != '\\')
+            {
+                //String ending found
+                stateStack.Pop();
+                objStack.Peek().AddProperty(currentName, new KVValue(typeof(String), currentString.ToString()));
+                return;
+            }
+
+            currentString.Append(c);
+        }
+
+        //Reading multiline string
+        private static void ReadValueStringMulti(char c)
+        {
+            //Check for ending
+            if (c == '"' && c == '"' && fileStream[index + 1] == '"' && fileStream[index + 2] == '"' && fileStream[index-1] != '\\')
+            {
+                stateStack.Pop();
+                objStack.Peek().AddProperty(currentName, new KVValue(typeof(String), currentString.ToString()));
+
+                //Skip to end
+                index += 2;
+                return;
+            }
+
+            currentString.Append(c);
+        }
+
+        //Read a boolean variable
+        private static void ReadValueBool(char c)
+        {
+            //Stop reading once the end of true or false is reached
+            if (c == 'e')
+            {
+                currentString.Append(c);
+                stateStack.Pop();
+                objStack.Peek().AddProperty(currentName, new KVValue(typeof(Boolean), currentString.ToString() == "true" ? true : false));
+                return;
+            }
+
+            currentString.Append(c);
+        }
+
+        //Read a numerical value
+        private static void ReadValueNumber(char c)
+        {
+            //For arrays
+            if (c == ',')
+            {
+                stateStack.Pop();
+                stateStack.Push(State.SEEK_VALUE);
+                if ( currentString.ToString().Contains('.')) {
+                    objStack.Peek().AddProperty(currentName, new KVValue(typeof(Double), Double.Parse(currentString.ToString(), CultureInfo.InvariantCulture)));
+                } else {
+                    objStack.Peek().AddProperty(currentName, new KVValue(typeof(Int32), Int32.Parse(currentString.ToString(), CultureInfo.InvariantCulture)));
```

Comment:
Switched to a KVType enum instead of doing typeof stuff each time for the next commit.